### DISCUSSION TITLE
[Proposal]: Make C# as a native language without CLR

### DISCUSSION THREAD

**Discussion by taozuhong** [dotnet/csharplang#7150] [General]
# Native compiler support

* [x] Proposed
* [ ] Prototype: Not Started
* [ ] Implementation: Not Started
* [ ] Specification: Not Started

## Summary
Make a C# compiler like C/C++ without Managed runtime support

## Motivation
[motivation]: #motivation
I do like C# language, but it has a huge managed runtime, I do hope C# language has native compiler support.

Now I'm using C# native version: [Vala](https://gitlab.gnome.org/GNOME/vala)

**Comment by huoyaoyuan** [comment]
Does [NativeAOT](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/) meet your requirement? There are even someone using it to write UEFI firmware (of course not using the BCL).

**Reply by taozuhong** [reply]
Yet meet the requirement, the [NativeAOT](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/#limitations-of-native-aot-deployment) has many limitations.

**Reply by huoyaoyuan** [reply]
What are the specific limitations? A lot of limitations are bound tightly to the type system model used by C#.

**Comment by HaloFour** [comment]
I don't believe the language specification takes an opinion on subjects like this.  That'd be up to the implementation of the compiler.

**Comment by taozuhong** [comment]
Kotlin has a native compiler, Go/Rust/Swift are native compilers too, maybe C# too.

**Reply by theunrepentantgeek** [reply]
What's not sufficient about [Single file deployment](https://learn.microsoft.com/en-us/dotnet/core/deploying/single-file/overview?tabs=cli) & [Native Compilation](https://learn.microsoft.com/en-us/dotnet/core/deploying/native-aot/?tabs=net7).

**Reply by HaloFour** [reply]
This is not the repository to discuss a native compiler.  The language specification of C# doesn't care about whether it is natively compiled or not.

Microsoft already has projects to support native execution on various platforms, such as NativeAOT and Xamarin.
